cmd/athenz-sia/main.go

```diff
@@ -75,41 +75,93 @@ func main() {
 	}
 
 	// re-init logger from user config
-	log.InitLogger(filepath.Join(idConfig.LogDir, fmt.Sprintf("%s.%s.log", serviceName, idConfig.LogLevel)), idConfig.LogLevel, true)
-	log.Infof("Starting [%s] with version [%s], built on [%s]", APP_NAME, VERSION, BUILD_DATE)
-	log.Infof("Booting up with args: %v, config: %+v", os.Args, idConfig)
+	log.InitLogger(filepath.Join(idCfg.LogDir, fmt.Sprintf("%s.%s.log", serviceName, idCfg.LogLevel)), idCfg.LogLevel, true)
+	log.Infof("Starting [%s] with version [%s], built on [%s]", version.APP_NAME, version.VERSION, version.BUILD_DATE)
+	log.Infof("Booting up with args: %v, config: %+v", os.Args, idCfg)
+
+	// delay boot with jitter
+	if idCfg.DelayJitterSeconds != 0 {
+		sleep := time.Duration(rand.Int63n(idCfg.DelayJitterSeconds)) * time.Second
+		log.Infof("Delaying boot with jitter [%s] randomized from [%s]...", sleep, time.Duration(idCfg.DelayJitterSeconds)*time.Second)
+		time.Sleep(sleep)
+	}
+
+	// register metrics
+	metrics.RegisterBuildInfo(filepath.Base(os.Args[0]), version.VERSION, version.BUILD_DATE)
+
+	// variables
+	causeBySignal := fmt.Errorf("received signal")
+	causeByStartFailed := fmt.Errorf("start failed")
+	initCtx, cancelInit := context.WithCancelCause(context.Background())
+	runCtx, cancelRun := context.WithCancelCause(context.Background())
 
-	certificateChan := make(chan struct{}, 1)
-	ch := make(chan os.Signal, 1)
-	signal.Notify(ch, syscall.SIGTERM, os.Interrupt)
+	// signal handling in background
+	sigChan := make(chan os.Signal, 1)
+	signal.Notify(sigChan, syscall.SIGTERM, os.Interrupt)
+	go func() {
+		// wait for a signal, then cancel contexts
+		sigGot := <-sigChan
+		log.Infof("Received signal: %s", sigGot.String())
+		cancelInit(fmt.Errorf("%w: %s", causeBySignal, sigGot))
+		cancelRun(fmt.Errorf("%w: %s", causeBySignal, sigGot))
+	}()
 
-	err, sdChan := identity.Certificated(idConfig, certificateChan)
+	// initiate background services
+	certService, err := certificate.New(initCtx, idCfg)
 	if err != nil {
-		log.Fatalln(err)
+		log.Fatalf("Error initiating certificate provider: %s", err.Error())
+	}
+	tokenService, err := token.New(initCtx, idCfg)
```

Comment:
@WindzCUHK Is there any particular reason of changing the naming style from using the word "daemon" to using the word "service"?
I see the usage is quite similar and if the purpose is to just inherit the naming style that existed in Athenz/athenz-client-sidecar, then I don't believe the reason is sufficient to break the consistency.

Reply:
It is my personal understand. Please correct me.
- daemon implies something running in background.
- service provide some feature for the end user.
- From the user's point of view, the token service provides token as a service.
- From the developer's point of view, the token service is able to run in background so it should implements the daemon interface.
- The main focus should be providing some services to the end user. Hence, it should be service rather than daemon. (if it should be called as daemon service, simplify as service for the name.)

Reply:
extra review

Reply:
@WindzCUHK Just to clarify, but I'm not actually against this change. I'm only concerning about the consistency.

Reply:
filename: `daemon.go` => `service.go`